docs/specs/alcproxies.md

```diff
@@ -0,0 +1,251 @@
+	ProxyBuilder p = new ProxyBuilder(typeof(SampleInterface), alc, assemblyPath);
+	SampleInterface testObject = (SampleInterface)p.CreateObject("SampleType");
+	testObject.SampleMethod();
+	...
+
+}
+```
+The API surface allows for developers to create the ProxyBuilder object, and then create an instance of the proxy type, which works exactly like a normal instance of the target type, with the limitation that the target type must inherit from an interface that can be used to build the proxy object. This does limit the use of the object slightly, as use of the target object is limited to methods that exist in the interface.
+
+---
+
+### General Design
+
+![](../img/ALCProxyDesign.svg)
+
+The ProxyBuilder API will generate five separate objects:
+* `ProxyBuilder`: Responsible for creating `ProxyObject`, `ClientObject`, and `DispatchBuilder` and returns the instance of the desired proxy to the user for use.
+* `DispatchBuilder`: Constructs `ServerObject` in the target ALC.
+* `ProxyObject`: The object that the user has access to, it allows for method calls to run on the `TargetObject` in the target ALC. When a method is called, `ProxyObject` pipes its command to the `ServerObject`, which will eventually from the user context to the new proxy now in the target ALC using binary serialization.
+* `ClientObject`: Sends messages to the `ServerObject` in the target ALC requesting `TargetObject`'s method to be called.
+* `ServerObject`: Invokes `TargetObject`'s methods, and pipes the output back to `ClientObject` and the user ALC.
+
+When a method is called, `ProxyObject` pipes its command to the `ClientObject`, which will serialize or encode the method information and any parameters passed to the `ProxyObject`, and send them to `ServerObject` through some sort of request. `ServerObject` then deserializes the data, and uses reflection to invoke the method in `TargetObject`. Once the method runs and `ServerObject` gets the returned object, it sends it back to `ClientObject`, which sends it back to `ProxyObject` to be "returned" to the user.
+
+A `DispatchProxy` can be used to create a "fake" object of a given type, which acts as if it was the instance we wanted to use. Having `ProxyObject` inherit from `DispatchProxy` allows us to easily add more extra actions when we call the object, such as deserialization of objects across the type barrier (See the "Type-Isolation Barrier" section below).
+
+This design is dependent on the ability for the `ProxyBuilder` to create dynamic assemblies and types within a different ALC/Process. Currently we may be able to do so by loading the `DispatchBuilder` type into the ALC directly, and then creating the proxy object from there, while also using a similar builder to create the user-side proxy that make gRPC calls. We can also wait (See [#38426](https://github.com/dotnet/corefx/issues/38426 )), but this option will not be available until said issue can be fixed, or a comparable fix is made.
```

Comment:
It's more that we want to create the server in the target ALC so it can easily know the exact types that the `TargetObject` needs to use (its own type, its interface, the arguments, etc.) and do any sort of decoding/deserialization from the client correctly. It may be possible to do all that through the client, or anywhere on the user ALC side, but it'll take a lot of reflection and weird shenanigans to get that working, since to the user ALC, those types don't technically exist. I think if we used more shared types (like what you are mentioning in other comments) it works, but not if we're assuming that stuff may not be shared.

Reply:
> to the user ALC, those types don't technically exist

I wouldn't say that statement is accurate. They are exist, but are not the same type.

There is no type isolation boundary.  There is a type resolution boundary. The type is loaded in exactly one ALC. 

Each type can be referenced by multiple ALCs.

https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.getassemblies?view=netcore-3.0 Will return all the assemblies in the entire process.

https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.gettypes?view=netcore-3.0 Will return all the types in an Assembly.

There is no isolation.

You can easily iterate through all types in the process.

You can also get any specific type from any specific AssemblyLoadContext using https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.gettype?view=netcore-3.0 

It is a bit more difficult to get a specific ALCs view of a given type name by string.  This https://docs.microsoft.com/en-us/dotnet/api/system.type.gettype?view=netcore-3.0#System_Type_GetType_System_String_System_Func_System_Reflection_AssemblyName_System_Reflection_Assembly__System_Func_System_Reflection_Assembly_System_String_System_Boolean_System_Type__ is the general solution.  You need to provide a function which will ask a specific ALC an Assembly by name. The type resolver can be `null`.